### DISCUSSION TITLE
Installing ACPP on laptop with integrated GPU (Intel Core 7 Ultra 155U  )

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1626] [General]
Hello,

I wonder if someone could help me. This is more an installation issues.

I have a laptop with integrated GPU and I managed to install OneAPI and make codes to run on both cpu and gpu. I have Ubuntu 24 running in wsl2, but I took the instructions from this page https://www.intel.com/content/www/us/en/docs/oneapi/installation-guide-linux/2024-0/configure-wsl-2-for-gpu-workflows.html (corresponding to Ubuntu 22). I managed to run different codes without issues. 
I then compiled Acpp with just the options `cmake -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_C_COMPILER=clang-$LLVM_VERSION -DCMAKE_CXX_COMPILER=clang++-$LLVM_VERSION -DACPP_COMPILER_FEATURE_PROFILE=full`. 

This worked well, but when I ran `accp-info` it only showed the cpu as target. 
Then I tried the option `-DWITH_OPENCL_BACKEND=ON`. During the configuration step  it complained that it could not find OpenCL. 

I made a search on the computer, but the only `libOpenCL.so` was found in the OneAPI. I indicated this in the configuration, but this one was not recognized. 

In the end after some search I found that there is a package on ubuntu roughly named ocl-opencl and install it. This created a library in `usr/local/lib` .

I recompile ACPP and this is the output of `acpp-info` looked like this:

```
$ acpp-info
=================Backend information===================
Loaded backend 0: OpenCL
  Found device: Intel(R) Core(TM) Ultra 7 155H
  Found device: Intel(R) Graphics [0x7d55]
Loaded backend 1: OpenMP
  Found device: hipSYCL OpenMP host device

=================Device information===================
***************** Devices for backend OpenCL *****************
Device 0:
 General device information:
  Name: Intel(R) Core(TM) Ultra 7 155H
  Backend: OpenCL
  Vendor: Intel(R) Core(TM) Ultra 7 155H
  Arch: OpenCL OpenCL C 3.0
  Driver version: 2024.18.10.0.08_160000
  Is CPU: 1
  Is GPU: 0
 ....
 Device 1:
 General device information:
  Name: Intel(R) Graphics [0x7d55]
  Backend: OpenCL
  Vendor: Intel(R) Graphics [0x7d55]
  Arch: OpenCL OpenCL C 1.2
  Driver version: 23.43.027642
  Is CPU: 0
  Is GPU: 1
```

I was able to compile sycl codes with `acpp-tagets=generic` and able to run simple codes like device info query and simple vector addition, but when I ran something with profiling and events I get this errorr:

```
 $ acpp -fuse-ld=lld -O3 --acpp-targets=generic j_simple_with_buffer.cpp
$ ./a.out
Offload Device        : Intel(R) Graphics [0x7d55]
max_work_group_size   : 1024
Configuration         : MATRIX_SIZE= 1024x1024
 [0][0] = 512
Warm up the device  
[AdaptiveCpp Warning] kernel_cache: This application run has resulted in new binaries being JIT-compiled. This indicates that the runtime optimization process has not yet reached peak performance. You may want to run the application again until this warning no longer appears to achieve optimal performance.
terminate called after throwing an instance of 'hipsycl::sycl::exception'
  what():  Operation not profiled: No execution end timestamp available
Aborted (core dumped)
```
This is a snip of the code that I think it is not supported by my installation:
```
    sycl::property_list q_prof{property::queue::enable_profiling{}, sycl::property::queue::in_order{}}; 
    queue q{default_selector_v,q_prof};
for(int iter=0;iter<niter; iter++)
    {       
        {
            //# Create buffers for matrices
            buffer<float, 1> u(matrix_u.data(), range<1>(nx*ny));
            buffer<float, 1> unew(matrix_unew.data(), range<1>(nx*ny));
            
            //# Submit command groups to execute on device            
            e = q.submit([&](handler &h){
                //# Create accessors to copy buffers to the device
                auto U = u.get_access<access::mode::read>(h);
                auto UNEW= unew.get_access<access::mode::write>(h);
                
                //# Define size for ND-Range and work-group size
                range<2> global_size(nx,ny);
                range<2> work_group_size(M,M);
                
                //# Parallel Compute 
                h.parallel_for(nd_range<2>{global_size, work_group_size}, [=](nd_item<2> item){
                    const int i = item.get_global_id(0);
                    const int j = item.get_global_id(1);
                    int ind = i * ny + j;
                    int ip = (i + 1) * ny + j;
                    int im = (i - 1) * ny + j;
                    int jp = i * ny + j + 1;
                    int jm = i * ny + j - 1;
                    if(i>0 && i<nx-1 && j>0 && j< ny-1){
                        UNEW[ind] = factor * (U[ip] - 2.0 * U[ind] + U[im] +
                                 U[jp] - 2.0 * U[ind] + U[jm]);
                    } 
        
                });
           });
        }
        kernel_duration += (e.get_profiling_info<info::event_profiling::command_end>() - e.get_profiling_info<info::event_profiling::command_start>());
       ```

My guess is that I installed OpenCL bad. Can anyone give a hint about how to install OpenCL for ACPP?

Cristian

**Comment by illuhad** [comment]
Your program is working correctly. Kernel profiling and event timestamps are an optional feature that is not supported by the AdaptiveCpp OpenCL backend. Applications must query availability of this feature using `device::has(aspect::queue_profiling)` prior to using it.

**Reply by csccva** [reply]
Thanks for your reply. That's good to know. 

Cristian

**Reply by csccva** [reply]
I forgot to ask. Is it possible to use level-zero? Would something change?

**Reply by illuhad** [reply]
The Level Zero backend does not support this feature either. It also is generally less stable than the OpenCL backend and less optimized, so I would not recommend using it for production workloads.

Both the OpenCL and Level Zero APIs share the design that profiling is enabled/disabled on a per-queue basis, whereas AdaptiveCpp wants to enable/disable it on a per-kernel basis due to the AdaptiveCpp's runtime design relying on global queue pools. This makes it difficult to implement for these two backends. All other backends support queue profiling.

**Reply by csccva** [reply]
I understand. Thanks. 
I can stick with OneAPI, but I want try during development to keep the code as much as possible "compilable" with ACPP for non-intel machines. It is just kernels and blas call. I do not want to have 3-4 versions of the same code. Well apart for the blas part, of course. 

Cristian
